вопрос

KloopRE opened the pull request:
Почему по нажатию кнопок управления ничего не происходит?  возможно, что-то блокирует обработку событий клавиш

KloopRE committed:
вопрос

Почему по нажатию кнопок управления ничего не происходит?  возможно, что-то блокирует обработку событий клавиш

hBuzzy reviewed (commented):
У вас фокус слетает на другой элемент. Установите просто еще одну строку:
```
setFocusPolicy(Qt::StrongFocus);
setFocus();
```

Кроме того, задание подразумевает (как описано в разделе интерфейс), использование `QGraphicsView` и, соответственно, `QGraphicsScene` для отображения поля. Другие варианты в задаче не принимаются.

KloopRE committed:
Помогите

В чём ошибка?

D:\Pacman-task\hostile.cpp:143: ошибка: undefined reference to `Hostile::Hostile(std::vector<std::vector<int, std::allocator<int> >, std::allocator<std::vector<int, std::allocator<int> > > >&)'

D:\Pacman-task\mainwindow.cpp:90: ошибка: undefined reference to `Hostile::Hostile()'

hBuzzy reviewed (commented):
У вас в hostile.h определены два конструктора, а где их реализация в  hostile.cpp? Что вызываться то должно при создании объекта? На это вам компилятор и ругается.

+ Подумайте, точно ли position_ должен быть ссылкой?

KloopRE committed:
Еще вопрос

У меня проблема с событием перемещения мыши. Когда я вожу по краям своей формы ui, то событие срабатывает и выводятся координаты, но когда я вожу по его внутренней составляющей он не срабатывает.
Я думаю, что проблема в этих строках:
scene = new QGraphicsScene(this);
    view = new QGraphicsView(scene, this);

    view->setMouseTracking(true);

Можете помочь?

KloopRE commented:
Последний комит относится к разделу редактирования игровой сетки

hBuzzy reviewed (commented):
У вас есть главный виджет для которого вы переопределили `mouseMoveEvent` и на этот виджет вы добавляете `QGraphicsView`. Как только вы переводите мышь на view, главный виджет теряет фокус и у него больше не вызывается `mouseMoveEvent`. Отсюда и ничего не выводится. Если вам нужен этот функционал, переопределяйте передвижение мыши для других виджетов.

KloopRE committed:
redactor

add redactor

KloopRE committed:
pacman

edir comments

hBuzzy commented on customgraphicsview.cpp:
```diff
@@ -0,0 +1,38 @@
+
+CustomGraphicsView::CustomGraphicsView(QGraphicsScene *scene, QWidget *parent, int gridSize)
+    : QGraphicsView(scene, parent), gridSize(gridSize)  // Добавьте инициализацию gridSize
+{
```
Раз начали не переносить скобки в части проекта, то не переносите их везде. Касается всего проекта. + Комментарий.

hBuzzy commented on customgraphicsview.cpp:
```diff
@@ -0,0 +1,38 @@
+
```
Лишняя пустая строка.

hBuzzy commented on customgraphicsview.cpp:
```diff
@@ -0,0 +1,38 @@
+        QPoint localMousePos = event->pos();
+        center = this->mapToScene(localMousePos);
+        center.setX(this->center.x() - gridSize / 2);
+        center.setY(this->center.y() - gridSize / 2);
```
2 - магическое число.

hBuzzy commented on customgraphicsview.h:
```diff
@@ -0,0 +1,27 @@
+public:
+    CustomGraphicsView(QGraphicsScene *scene, QWidget *parent = nullptr, int gridSize = 64);
+    QPointF center;
+    bool isdrawing = false;
```
Публичных полей, если они не константы, быть не должно. Сделайте их приватными, доступ другими классами к полям осуществляется через методы.

hBuzzy commented on customgraphicsview.h:
```diff
@@ -0,0 +1,27 @@
+    void mouseReleaseEvent(QMouseEvent *event) override;
+
+private:
+    int gridSize;
```
К приватным полям добавляется постфикс "_". `gridSize` -> `gridSize_`

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+using namespace std;
+
+
+
```
Не больше одной пусто строки

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+
+Hostile::Hostile() : position_(0, 0) {
+
+}
```
В данном случае, для экономии места, можно скобки написать на одном уровне, без переноса закрывающей.

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+
+bool Hostile::isValid(int x, int y) {
+    int rows = matrix_.size();
+    int cols = matrix_[0].size();
```
col с английского - седло. Не сокращайте названия.

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+bool Hostile::isValid(int x, int y) {
+    int rows = matrix_.size();
+    int cols = matrix_[0].size();
+    return x >= 0 && x < cols && y >= 0 && y < rows && matrix_[y][x] != 1 && matrix_[y][x] != 3;
```
магические числа

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+        directions.push_back('D');  // Вниз
+    } else if (dy < 0) {
+        directions.push_back('U');  // Вверх
+    }
```
R L D U - Это? 

Во-первых, т.к. это используется в разных частях кода, то выделите направления в константы и дайте нормальные имена и значения. Во-вторых, комментарии - удалить.

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+
+vector<char> Hostile::getPath(const Point& end) {
+    int rows = matrix_.size();
+    int cols = matrix_[0].size();
```
Без сокращений

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+
+    vector<vector<int>> distance(rows, vector<int>(cols, INT_MAX));
+    vector<vector<Point>> path(rows, vector<Point>(cols, Point(-1, -1)));
+    priority_queue<Node, vector<Node>, greater<Node>> pq;
```
Без сокращений

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+            return directions;
+        }
+
+        for (int i = 0; i < 4; ++i) {
```
4 - магическое число

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+    }
+};
+
+bool Hostile::isValid(int x, int y) {
```
Что конкретно проверяется на валидность? Отразите в названии.

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+            int nextY = current.point.y + dy[i];
+
+            if (isValid(nextX, nextY)) {
+                int newDistance = current.distance + 1; // Вес каждого шага равен 1
```
Комментарий

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+Point Hostile::getPosition()
+{
+    return position_;
+}
```
Опять же, везде один стиль скобок, раз начали не переносить, то не переносите везде.

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+
+void Hostile::setMatrix(std::vector<std::vector<int> > &inputMatrix)
+{
+    matrix_ = inputMatrix;
```
Input, в целом, бесполезный придаток. Чтобы не было конфликта имен у нас есть у полей постфикс.

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+    return position_;
+}
+
+int Primer() {
```
Не соблюдена нотация названия методов. Если это должен быть класс - выделите, иначе соблюдайте нотацию методов / функций.

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,169 @@
+        {1, 0, 0, 0, 0}
+    };
+
+    Hostile hostileObject(inputMatrix);
```
Object в данном контексте бесполезен.

hBuzzy commented on hostile.h:
```diff
@@ -0,0 +1,36 @@
+    int x, y;
+
+    Point(int _x, int _y) : x(_x), y(_y) {}
+};
```
Чем вам не угодила QPoint или QPointF ? Зачем создавать костыли?

hBuzzy commented on hostile.h:
```diff
@@ -0,0 +1,36 @@
+    void setPosition(Point& pos);
+    void setMatrix(std::vector<std::vector<int>>& inputMatrix);
+    Point getPosition();
+    int temp = 0;
```
Сначала все public, потом protected и только потом private. 

При этом, внутри сортировки по доступу, сначала конструкторы, потом методы, потом поля. Подробнее в теор. материалах.

hBuzzy replied:
temp - не общепринятое сокращение. Плюс временный в данном контексте смотрится крайне плохо. Временное что? К тому же, без public полей.  Кроме того, не вижу, чтобы оно где-то использовалось.

hBuzzy commented on mainwindow.h:
```diff
@@ -1,21 +1,63 @@
+private slots:
+    void updateGameTime();
+    void updateHostileRunTime();
+    void handleKeyRepeat(); 
```
1. Наладьте порядок в классе. Прочитайте тоер материал.
2. Константы названы неверно
3. Не мешайте переменные с методами
4. handleKeyRepeat нигде не используется

hBuzzy commented on mainwindow.h:
```diff
@@ -1,21 +1,63 @@
+    float gameTime_;
+    QTimer* hostileRunTimer_;
+    const int rows_ = 10;
+    const int cols_ = 10;
```
Без сокращений

hBuzzy commented on mainwindow.h:
```diff
@@ -1,21 +1,63 @@
+    const int cols_ = 10;
+    int **gameGrid_;
+    int gridSize_ = 64;
+    int countCoins_ = 5;
```
`countCoins_ `-> `counsCount_`

hBuzzy commented on mainwindow.h:
```diff
@@ -1,21 +1,63 @@
+    void movePlayerRight();
+    void moveHostile();
+    QGraphicsScene *scene;
+    QGraphicsView *view;
```
У приватных полей постфикс "_"

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
-      , ui(new Ui::MainWindow)
+    : QMainWindow(parent), ui(new Ui::MainWindow) {
+
+
```
Не больше одной пустой строки.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+    view = new QGraphicsView(scene, this);
+
+    int uiWidth = (cols_+4) * gridSize_;
+    int uiHeight = (rows_+1) * gridSize_;
```
Пробелы вокруг мат. операторов и в добавок магические числа.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+
+    gameTimer_ = new QTimer(this);
+    gameTimer_->start(250);
+    connect(gameTimer_, SIGNAL(timeout()), this, SLOT(updateGameTime()));
```
Не используем старую запись соединения, перепишите на новую.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+MainWindow::MainWindow(QWidget *parent, int** gameGrid)
+    : QMainWindow(parent), ui(new Ui::MainWindow) {
+
+
```
Не больше одной пустой строки

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+    ui->setupUi(this);
+    setFocusPolicy(Qt::StrongFocus);
+
+    hostiles_.reserve(cols_*rows_);
```
Пробелы

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+    view = new QGraphicsView(scene, this);
+
+    int uiWidth = (cols_+4) * gridSize_;
+    int uiHeight = (rows_+1) * gridSize_;
```
Часть конструктора - дубляж кода из предыдущего. Вынесите общую логику в отдельный метод.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+        generateRandomGameGrid();
+    }
+
+    setupGameGrid();
```
Вообще, такие огромные методы / конструкторы и т.д. явно напрашиваются на разбивку на разные методы. Не создавайте таких монстров, если можно делить логику - делите.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+            }
+    }
+    Hostile newHostile;
+    Point p(0, 0);
```
p?

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+void MainWindow::gameOver(bool isWin)
 {
-  ui->setupUi(this);
+    disconnect(gameTimer_, SIGNAL(timeout()), this, SLOT(updateGameTime()));
+    disconnect(hostileRunTimer_, SIGNAL(timeout()), this, SLOT(updateHostileRunTime()));
```
Используйте новую запись, вместо старой.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
-  ui->setupUi(this);
+    disconnect(gameTimer_, SIGNAL(timeout()), this, SLOT(updateGameTime()));
+    disconnect(hostileRunTimer_, SIGNAL(timeout()), this, SLOT(updateHostileRunTime()));
+    if(isWin)
```
пробел + скобки без переноса.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+    else
+    {
+        std::string str1 = "Game Over, coins left: ";
+        std::string str2 = std::to_string(countCoins_);
```
str1 и str2 ?

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+    disconnect(hostileRunTimer_, SIGNAL(timeout()), this, SLOT(updateHostileRunTime()));
+    if(isWin)
+    {
+        std::string str1 = "You WIN!";
```
str1 ? Почему она тут 1? Дайте нормальное название.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+    }
+
+    int randomRow = std::rand() % rows_;
+    int randomCol = std::rand() % cols_;
```
Без сокращений

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+    generateRandomElements(element,count);
+
+    element = 4;
+    generateRandomElements(element,countCoins_);
```
Пробелы после запятой. Учитывая. что вы передаете разные параметры в функцию и у параметров нормальные имена, можно было не выделять значения в переменные.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+
+void MainWindow::generateRandomElements(int element, int count) {
+    Hostile newHostile;
+    Point p(0,0);
```
p?

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+void MainWindow::generateRandomElements(int element, int count) {
+    Hostile newHostile;
+    Point p(0,0);
+    if(element == 3){
```
Магическое значение

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+                    std::cout << element << " ";
+                }
+                std::cout << std::endl;
+            }
```
Выводы массивов и т.д. отлично выносятся в отдельные методы не забивая основную логику происходящего.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+    }
+    for (int k = 0; k < count; ++k) {
+        int randomRow = std::rand() % rows_;
+        int randomCol = std::rand() % cols_;
```
Без сокращений

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+            }
+            newHostile.setMatrix(matrix);
+    }
+    for (int k = 0; k < count; ++k) {
```
В чем смысл k, если область видимости i уже доступна? Если это не вложенные циклы, подобного не требуется.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+        randomRow = std::rand() % rows_;
+        randomCol = std::rand() % cols_;
+        }
+        if(element == 3)
```
Пробел, магическое число, скобки не переносим. Неплохо было бы еще и пустую строку добавить перед if

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+                    coinItem->setX(j * gridSize_);
+                    coinItem->setY(i * gridSize_);
+                    break;
+                }
```
Уже само то, что чтобы понять что мы получаем под числами, необходимо комментировать, означает, что код написан плохо. В идеале в case должны быть либо константы, либо перечисления, но никак не подобное. Добавьте просто хотя бы int константы и сравнивайте с ними.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+            for (int j = 0; j < cols_; ++j) {
+                switch (gameGrid_[i][j]) {
+                case 0: // Empty space
+                    scene->addRect(j * gridSize_, i * gridSize_, gridSize_, gridSize_, QPen(Qt::black), QBrush(Qt::white));
```
Больше похоже, что gridSize у вас это cellSize, если я не ошибаюсь.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+            gameGrid_[hostilePosition.y][hostilePosition.x] = 3;
+            qDebug() << "gameGrid_ Coordinates: (" << hostilePosition.x << ", " << hostilePosition.y << ")";
+            currentHostile.setPosition(hostilePosition);
+        }
```
Явный дубляж кода. Вынесите общую логику отдельно. 3 - магическое число. Не везде есть пробелы.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+
+void MainWindow::keyPressEvent(QKeyEvent *event) {
+    if(isKeyTime_)
+    {
```
Пробел после if, скобку не переносим

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,464 @@
+            qDebug() << "Player Coordinates: (" << player_.getY() << ", " << player_.getX() << ")";
+        }
+    }
+}
```
Все 4 методы движения - явный дубляж кода. Выделите общую логику отдельно. Не везде есть пробелы. Магические числа.

hBuzzy commented on player.h:
```diff
@@ -0,0 +1,20 @@
+    int getX() const;
+    int getY() const;
+    void setX(int newX);
+    void setY(int newY);
```
Исправить последовательность в классе. Убрать комментарии.

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+Redactor::Redactor(QWidget *parent) :
+    QMainWindow(parent),
+    ui(new Ui::Redactor)
+{
```
Не переносим скобку

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+    connect(view, &CustomGraphicsView::customMouseRelease, this, &Redactor::handleCustomMouseRelease);
+
+    int uiWidth = (cols_+6) * gridSize_;
+    int uiHeight = (rows_+1) * gridSize_;
```
Уже писал замечания к подобным строка - тут аналогично.

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+    }
+
+    myButton = new QPushButton("Начать игру", this);
+    myButton->setGeometry(10, 10, 100, 30); // Установите положение и размер кнопки
```
Комментарий

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+}
+
+void Redactor::exitRedaction()
+{
```
Не переносим скобку, исправьте во ВСЕМ проекте.

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+}
+
+void Redactor::openGame() {
+    MainWindow *SG = new MainWindow(nullptr, gameGrid_);
```
SG?

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+        delete dragItem_;
+    }
+    dragItem_ = new QGraphicsPixmapItem();
+    if(dragElement_ == 1)
+    {
+        dragItem_ = scene->addPixmap(QPixmap(":resource/wall.png").scaled(gridSize_, gridSize_));
+    }
+    if(dragElement_ == 2)
+    {
+        dragItem_ = scene->addPixmap(QPixmap(":resource/puckman.png").scaled(gridSize_, gridSize_));
+    }
+    if(dragElement_ == 3)
+    {
+        dragItem_ = scene->addPixmap(QPixmap(":resource/hostile.png").scaled(gridSize_, gridSize_));
+    }
+    if(dragElement_ == 4)
+    {
+        dragItem_ = scene->addPixmap(QPixmap(":resource/coin.png").scaled(gridSize_, gridSize_));
+    }
```
1 ... 4 - это что? Пробелы после if

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+
+void Redactor::setupGameGrid() {
+        scene->clear();
+        QGraphicsPixmapItem* puckmanItem;
+        QGraphicsPixmapItem* hostileItem;
+        QGraphicsPixmapItem* coinItem;
```
Item особо не имеет смысла, по основным названиям и типу данных будет и так понятно.

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+        for (int i = 0; i < rows_; ++i) {
+            for (int j = 0; j < cols_; ++j) {
+                switch (gameGrid_[i][j]) {
+                case 0: // Empty space
+                    scene->addRect(j * gridSize_, i * gridSize_, gridSize_, gridSize_, QPen(Qt::black), QBrush(Qt::white));
+                    break;
+                case 1: // Wall
+                    scene->addRect(j * gridSize_, i * gridSize_, gridSize_, gridSize_, QPen(Qt::black), QBrush(Qt::darkGray));
+                    break;
+                case 2: // player
+                    puckmanItem = scene->addPixmap(QPixmap(":resource/puckman.png").scaled(gridSize_, gridSize_));
+                    puckmanItem->setX(j * gridSize_);
+                    puckmanItem->setY(i * gridSize_);
+                    break;
+                case 3: // hostile
+                    hostileItem = scene->addPixmap(QPixmap(":resource/hostile.png").scaled(gridSize_, gridSize_));
+                    hostileItem->setX(j * gridSize_);
+                    hostileItem->setY(i * gridSize_);
+                    break;
+                case 4: // hostile
+                    coinItem = scene->addPixmap(QPixmap(":resource/coin.png").scaled(gridSize_, gridSize_));
+                    coinItem->setX(j * gridSize_);
+                    coinItem->setY(i * gridSize_);
+                    break;
+                }
```
Та же проблема с case 1..4 - это что? Как понять ? Комментарии придется удалить.

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+                }
+            }
+        }
+        int RETURN = 2;
```
camelCase

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+}
+
+QPoint Redactor::getGridPoint() {
+    QPoint a(-1, -1);
```
a - это?

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+                    qDebug() << "getGridPoint Coordinates: (" << a.x() << ", " << a.y() << ")";
+                    return(a);
+                }
+            }
```
2 - магическое число. Пробелы после if, двух сторон от * и /. Скобки не переносим.

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+
+int Redactor::getElement() {
+    QGraphicsItem* item = scene->itemAt(view->center.x(), view->center.y(), QTransform());
+    if(view->center.x() < wallItem_->x()+gridSize_ && view->center.x() > wallItem_->x()-gridSize_)
```
Пробелы

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+    QGraphicsItem* item = scene->itemAt(view->center.x(), view->center.y(), QTransform());
+    if(view->center.x() < wallItem_->x()+gridSize_ && view->center.x() > wallItem_->x()-gridSize_)
+    {
+        if(view->center.y() < wallItem_->y()+gridSize_ && view->center.y() > wallItem_->y()-gridSize_)
+        {
+            qDebug() << "you drag item: (" << item << ")";
+            return 1;
+        }
+    }
+    if(view->center.x() < puckmanItem_->x()+gridSize_ && view->center.x() > puckmanItem_->x()-gridSize_)
+    {
+        if(view->center.y() < puckmanItem_->y()+gridSize_ && view->center.y() > puckmanItem_->y()-gridSize_)
+        {
+            if(isStatePacmanElement_ == false)
+            {
+                qDebug() << "you drag item: (" << item << ")";
+                return 2;
+            }
+        }
+    }
+    if(view->center.x() < hostileItem_->x()+gridSize_ && view->center.x() > hostileItem_->x()-gridSize_)
+    {
+        if(view->center.y() < hostileItem_->y()+gridSize_ && view->center.y() > hostileItem_->y()-gridSize_)
+        {
+            qDebug() << "you drag item: (" << item << ")";
+            return 3;
+        }
+    }
+    if(view->center.x() < coinItem_->x()+gridSize_ && view->center.x() > coinItem_->x()-gridSize_)
+    {
+        if(view->center.y() < coinItem_->y()+gridSize_ && view->center.y() > coinItem_->y()-gridSize_)
+        {
+            qDebug() << "you drag item: (" << item << ")";
+            return 4;
+        }
+    }
```
Пробелы

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+    if (event->button() == Qt::LeftButton) {
+        dragItem_ = new QGraphicsPixmapItem();
+        view->isdrawing = true;
+        QPoint localMousePos = event->pos();  // Получаем локальные координаты мыши
```
Комментарий

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+void Redactor::handleCustomMouseRelease() {
+    dragItem_ = new QGraphicsPixmapItem();
+    QPoint a = getGridPoint();
+    if(a.x() > 0 || a.y() > 0)
```
пробел после if

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,264 @@
+    QPoint a = getGridPoint();
+    if(a.x() > 0 || a.y() > 0)
+    {
+        if (gameGrid_[a.y()][a.x()] == 2)
+        {
+            isStatePacmanElement_ = false;
+        }
+        gameGrid_[a.y()][a.x()] = dragElement_;
+        if(dragElement_ == 2)
+        {
+            isStatePacmanElement_ = true;
+        }
```
пробел после if. 2 - магическое число

hBuzzy commented on redactor.h:
```diff
@@ -0,0 +1,59 @@
+    void paintEvent(QPaintEvent *event) override;
+    void mousePressEvent(QMouseEvent *event) override;
+
+private:
+    Ui::Redactor *ui;
+    MainWindow *MW;
+    int **gameGrid_;
+    const int rows_ = 10;
+    const int cols_ = 10;
+    int gridSize_ = 64;
+    void setupGameGrid();
+    QGraphicsScene *scene;
+    CustomGraphicsView *view;
+    bool isdrawing_ = false;
+    QPointF center_;
+    int lineWidth_;
+    QColor lineColor_;
+    void drawItems(QPainter *painter, const QPoint &center);
+    int getElement();
+    QGraphicsPixmapItem* wallItem_;
+    QGraphicsPixmapItem* puckmanItem_;
+    QGraphicsPixmapItem* hostileItem_;
+    QGraphicsPixmapItem* coinItem_;
+    QGraphicsPixmapItem* dragItem_ = nullptr;
+    QPoint getGridPoint();
+    int dragElement_;
+    void dragItem();
+    void handleCustomMouseRelease();
+    bool isStatePacmanElement_ = false;
+    QPushButton *myButton;
+    void handleButtonClick();
+    void openGame();
+    void exitRedaction();
```
Наведите порядок в сортировке. Не все поля имеют постфикс. Константы названы неверно, должен быть префикс "к". Опять же, gridSize - это именно размер всей сетки или ее ячейки?

hBuzzy commented on startgame.cpp:
```diff
@@ -0,0 +1,36 @@
+#include "mainwindow.h"
+#include "redactor.h"
+
+StartGame::StartGame(QWidget *parent) :
+    QWidget(parent),
+    ui(new Ui::StartGame)
+{
+    ui->setupUi(this);
+    loadUI();
+}
+
+StartGame::~StartGame()
+{
+    delete ui;
+}
+
+void StartGame::loadUI(){
+    ui->play_button->setFont(QFont(default_font_family_, font_size_));
+    ui->redactor_button->setFont(QFont(default_font_family_, font_size_));
+}
+
+void StartGame::on_play_button_clicked()
+{
+    this->close();
+    MainWindow *game = new MainWindow(nullptr, nullptr);
+    game->show();
+}
+
+void StartGame::on_redactor_button_clicked()
+{
+    this->close();
+    Redactor *game = new Redactor();
+    game->show();
+}
```
Скобки не переносим

hBuzzy commented on startgame.cpp:
```diff
@@ -0,0 +1,36 @@
+    ui->redactor_button->setFont(QFont(default_font_family_, font_size_));
+}
+
+void StartGame::on_play_button_clicked()
+{
+    this->close();
+    MainWindow *game = new MainWindow(nullptr, nullptr);
+    game->show();
+}
+
+void StartGame::on_redactor_button_clicked()
```
Лучше в camelCase даже именование стандартных событий.

hBuzzy commented on startgame.h:
```diff
@@ -0,0 +1,29 @@
+
+private:
+    Ui::StartGame *ui;
+    static const int font_size_ = 15;
+    const QString default_font_family_ = "Default";
```
Константы названы неверно. kCamelCase.

hBuzzy reviewed (changes requested):
Исправить замечания. Нигде не вижу проверки на коллизии -> задание выполнено не полностью.

Есть возможность запустить игру сразу после открытия редактора, при этом выводится сообщение о победе -> Неверная логика. Должно быть либо предупреждение о том, что должно быть добавлено для начала, либо просто ошибка.

На самом деле, своей реализацией через сетку, в убили всю логику задания с QGraphicsView, но да ладно, если вам так удобно - пусть движение будет осуществляться так.

KloopRE committed:
Исправление замечаний

Простите, можно мне пожалуйста не делать collidingItems, просто у меня уже прописана вся логика и мне придется её переписывать, я невнимательно читал задание : ( . Извините!

hBuzzy commented on customgraphicsview.cpp:
```diff
@@ -1,38 +1,74 @@
+
+void CustomGraphicsView::setDrawing(bool newDrawing)
+{
+    isDrawing_ = newDrawing;
```
newDrawing должно быть аналогично в нотации bool переменных. У вас есть поле isDrawing_, напрашивается сразу присвоить ему значение из переменной isDrawing. Тут нам, чтобы не изобретать велосипед, помогает разница в нотации переменных и приватных полей.

hBuzzy commented on customgraphicsview.cpp:
```diff
@@ -1,38 +1,74 @@
-    if (isdrawing) {
+void CustomGraphicsView::setCursorStyle()
+{
+    if (isDrawing_)
+    {
         setCursor(Qt::CrossCursor);
     } else {
         setCursor(Qt::ArrowCursor);
     }
```
У else неверно расставлены скобки. Раз везде решили не переносить, то:

```C++
if () 
{
    //Код
}
else
{
    //Код
}
```

hBuzzy commented on hostile.cpp:
```diff
@@ -4,60 +4,54 @@ using namespace std;
-    vector<vector<Point>> path(rows, vector<Point>(cols, Point(-1, -1)));
+    vector<vector<int>> distance(rows, vector<int>(columns, INT_MAX));
+    vector<vector<Point>> path(rows, vector<Point>(columns, Point(-1, -1)));
     priority_queue<Node, vector<Node>, greater<Node>> pq;
```
Все еще не исправлено сокращение.

hBuzzy commented on hostile.cpp:
```diff
@@ -66,34 +60,42 @@ vector<char> Hostile::getPath(const Point& end) {
-                    directions.push_back('U');
-                } else if (dy == 1) {
-                    directions.push_back('D');
+                if (dx == -1)
+                {
+                    directions.push_back("Left");
+                } else if (dx == 1)
+                {
+                    directions.push_back("Right");
+                } else if (dy == -1)
+                {
+                    directions.push_back("Up");
+                } else if (dy == 1)
+                {
+                    directions.push_back("Down");
```
Опять же, во многих частях кода есть данные строки с направлениями. Если начнем их менять, попадем в ад.

Если все в пределах одного класса, то можно создать константы внутри класса:

const QString kLeft = "Left" и т.д.

Если они разбросаны по разным классам, то просто создайте абстрактный класс Directions в котором будут статические константы направлений.

hBuzzy commented on hostile.cpp:
```diff
@@ -103,14 +105,20 @@ vector<char> Hostile::getPath(const Point& end) {
         }
 
-        for (int i = 0; i < 4; ++i) {
+        int step_weight = 1;
```
step_weight -> stepWeight.

hBuzzy commented on hostile.cpp:
```diff
@@ -103,14 +105,20 @@ vector<char> Hostile::getPath(const Point& end) {
 
-        for (int i = 0; i < 4; ++i) {
+        int step_weight = 1;
+        static const int numDirections = 4;
```
num - не общепринятое сокращение. Зачем тут static? Если это должна быть константа, то нужно оформить в верной нотации:

const int kDirectionsNumber = 4;

hBuzzy commented on hostile.cpp:
```diff
@@ -119,51 +127,65 @@ vector<char> Hostile::getPath(const Point& end) {
+        if (pathDirections.empty())
+        {
             cout << "No path found." << endl;
         } else {
```
Скобки у else и само ключевое слово не на том уровне.

hBuzzy commented on hostile.h:
```diff
@@ -7,30 +6,50 @@
     Point position_;
+    int previousElement_  = 0;
 
 public:
```
Поместите все, что Public выше private.

1 Public
2 Protected
3 Private

hBuzzy commented on mainwindow.cpp:
```diff
@@ -13,20 +12,22 @@
 
-    int uiWidth = (cols_+4) * gridSize_;
-    int uiHeight = (rows_+1) * gridSize_;
+    int uiWidth = (kColumns_+4) * cellSize_;
+    int uiHeight = (kRows_+1) * cellSize_;
```
Пробелы и маг числа.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -36,33 +37,32 @@ MainWindow::MainWindow(QWidget *parent)
 
-    int uiWidth = (cols_+4) * gridSize_;
-    int uiHeight = (rows_+1) * gridSize_;
+    int uiWidth = (kColumns_+4) * cellSize_;
+    int uiHeight = (kRows_+1) * cellSize_;
```
Опять же - пробелы  +  маг числа.

Не исправлено замечание про общую логику в конструкторах и дубояж кода в них.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -116,37 +122,44 @@ MainWindow::MainWindow(QWidget *parent, int** gameGrid)
     gameTime_ += 0.5;
     isKeyTime_ = true;
     setupGameGrid();
     if(gameGrid_[player_.getY()][player_.getX()] == 3)
```
Пробел после if, 3 - маг число.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -208,162 +225,181 @@ void MainWindow::generateRandomGameGrid() {
+void MainWindow::generateRandomElements(int element, int count)
+{
     Hostile newHostile;
     Point p(0,0);
```
Не исправлено сокращение

hBuzzy commented on mainwindow.cpp:
```diff
@@ -208,162 +225,181 @@ void MainWindow::generateRandomGameGrid() {
     Hostile newHostile;
     Point p(0,0);
-    if(element == 3){
+    if(element == 3)
```
3 - это ?

hBuzzy commented on mainwindow.cpp:
```diff
@@ -208,162 +225,181 @@ void MainWindow::generateRandomGameGrid() {
+            randomRow = std::rand() % kRows_;
+            randomColumn = std::rand() % kColumns_;
         }
         if(element == 3)
```
Не исправлено замечание про маг число.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -208,162 +225,181 @@ void MainWindow::generateRandomGameGrid() {
-                switch (gameGrid_[i][j]) {
-                case 0: // Empty space
-                    scene->addRect(j * gridSize_, i * gridSize_, gridSize_, gridSize_, QPen(Qt::black), QBrush(Qt::white));
+        coinsLabel->setPos(kColumns_ * cellSize_ + 10, 10);
```
10 - это?

hBuzzy commented on mainwindow.cpp:
```diff
@@ -208,162 +225,181 @@ void MainWindow::generateRandomGameGrid() {
+    {
         currentHostile.setMatrix(matrix);
         direction = currentHostile.getPath(playerPoint);
         if(direction.empty())
```
Пробел после if

hBuzzy commented on mainwindow.cpp:
```diff
@@ -208,162 +225,181 @@ void MainWindow::generateRandomGameGrid() {
             return;
         }
-        if(direction[0] == 'L')
+        if(direction[0] == "Left")
```
Пробелы после if тут и ниже

hBuzzy commented on mainwindow.cpp:
```diff
@@ -208,162 +225,181 @@ void MainWindow::generateRandomGameGrid() {
+            currentHostile.setPreviousElement(gameGrid_[hostilePosition.y+1][hostilePosition.x]);
             hostilePosition.y = hostilePosition.y + 1;
             gameGrid_[hostilePosition.y][hostilePosition.x] = 3;
             qDebug() << "gameGrid_ Coordinates: (" << hostilePosition.x << ", " << hostilePosition.y << ")";
```
Не исправлено замечание про дубляж кода. Вынесите общую логику в отдельный метод. Все что меняется, меняется с помощью параметров метода. Не везде расставлены пробелы y+1. 3 - магическое число, которое в коде встречается крайне часто.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -399,64 +436,72 @@ void MainWindow::keyPressEvent(QKeyEvent *event) {
-        if(gameGrid_[player_.getY()+1][player_.getX()] != 1)
+void MainWindow::movePlayerDown()
+{
+    if (player_.getY()+1 < kRows_)
```
player_.getY()+1 Пробелы.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -399,64 +436,72 @@ void MainWindow::keyPressEvent(QKeyEvent *event) {
-void MainWindow::movePlayerDown() {
-    if (player_.getY()+1 < rows_) {
-        if(gameGrid_[player_.getY()+1][player_.getX()] != 1)
+void MainWindow::movePlayerDown()
```
Все методы движения так и остались дубляжом кода.

hBuzzy commented on redactor.cpp:
```diff
@@ -18,31 +19,34 @@ Redactor::Redactor(QWidget *parent) :
 
-    int uiWidth = (cols_+6) * gridSize_;
-    int uiHeight = (rows_+1) * gridSize_;
+    int uiWidth = (kColumns_+6) * cellSize_;
+    int uiHeight = (kRows_+1) * cellSize_;
```
Пробелы и магические числа.

hBuzzy commented on redactor.cpp:
```diff
@@ -56,186 +60,210 @@ void Redactor::exitRedaction()
                 }
             }
         }
         int RETURN = 2;
```
Не исправлено замечание про неверную нотацию именования переменной.

hBuzzy commented on redactor.cpp:
```diff
@@ -56,186 +60,210 @@ void Redactor::exitRedaction()
-        puckmanItem_ = scene->addPixmap(QPixmap(":resource/puckman.png").scaled(gridSize_, gridSize_));
-        puckmanItem_->setX((cols_+RETURN) * gridSize_);
-        puckmanItem_->setY(4 * gridSize_);
+        puckmanItem_ = scene->addPixmap(QPixmap(":resource/puckman.png").scaled(cellSize_, cellSize_));
+        puckmanItem_->setX((kColumns_+RETURN) * cellSize_);
+        puckmanItem_->setY(4 * cellSize_);
```
Пробелы и магические числа, как тут, так и в coinItem, 1, 4, 6, 8 и  т.д.

hBuzzy commented on redactor.cpp:
```diff
@@ -56,186 +60,210 @@ void Redactor::exitRedaction()
-            return 4;
+            return GameElement::Coin;
         }
     }
```
Вообще почти нигде не расставлены пробелы. Ни после if, ни вокруг мат. операторов, ни вокруг операторов сравнения.

hBuzzy commented on redactor.cpp:
```diff
@@ -244,20 +272,30 @@ void Redactor::mousePressEvent(QMouseEvent *event) {
         {
             isStatePacmanElement_ = true;
         }
+        if(dragElement_ == GameElement::Coin)
```
Пробелы перед if

hBuzzy commented on redactor.cpp:
```diff
@@ -244,20 +272,30 @@ void Redactor::mousePressEvent(QMouseEvent *event) {
-    QPoint a = getGridPoint();
-    if(a.x() > 0 || a.y() > 0)
+    QPoint cell = getGridPoint();
+    if(cell.x() > 0 || cell.y() > 0)
```
Пробел перед if

hBuzzy reviewed (changes requested):
Исправить замечания.

KloopRE committed:
Исправление ошибок

hBuzzy commented on hostile.cpp:
```diff
@@ -0,0 +1,195 @@
+    if (dx > 0)
+    {
+        directions.push_back(Directions(Directions::Right));
+    } else if (dx < 0)
```
else if нужно перенести

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,307 @@
+{
+    if (counsCount_ <= 0)
+    {
+        std::string str1 = "couns count < 0";
```
str1 - бесполезное название, дайте осмысленное название.

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,307 @@
+        wallItem_->setX((extraColumns) * cellSize_);
+        wallItem_->setY(nextRow);
+        qDebug() << "wallItem Coordinates: (" << wallItem_->x() << ", " << wallItem_->y() << ")";
+        nextRow = 4 * cellSize_;
+
+        puckmanItem_ = scene->addPixmap(QPixmap(":resource/puckman.png").scaled(cellSize_, cellSize_));
+        puckmanItem_->setX((extraColumns) * cellSize_);
+        puckmanItem_->setY(nextRow);
+        qDebug() << "puckmanItem Coordinates: (" << puckmanItem_->x() << ", " << puckmanItem_->y() << ")";
+        extraColumns = 2;
+        nextRow = 6 * cellSize_;
+
+        hostileItem_ = scene->addPixmap(QPixmap(":resource/hostile.png").scaled(cellSize_, cellSize_));
+        hostileItem_->setX((extraColumns) * cellSize_);
+        hostileItem_->setY(nextRow);
+        qDebug() << "hostileItem Coordinates: (" << hostileItem_->x() << ", " << hostileItem_->y() << ")";
+        extraColumns = 2;
+        nextRow = 8 * cellSize_;
```
Все еще 6, 8, 4, 1 встречаются в коде в виде просто чисел. Что это за числа!? Все это - магические значения, никто, кроме того, кто писал код, сразу не поймет что это такое. Дайте им названия в полях или переменных.

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,307 @@
+                }
+            }
+        }
+        int extraColumns = kColumns_+2;
```
Пробел с двух сторон от оператора сложения.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+
+    gameTimer_ = new QTimer(this);
+    gameTimer_->start(250);
+    connect(gameTimer_, SIGNAL(timeout()), this, SLOT(updateGameTime()));
```
Все еще старая запись сигналов и словом, вместо новой с указателями.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+
+    hostileRunTimer_ = new QTimer(this);
+    hostileRunTimer_->start(1000);
+    connect(hostileRunTimer_, SIGNAL(timeout()), this, SLOT(updateHostileRunTime()));
```
Все еще старая запись сигналов и словом, вместо новой с указателями.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+
+    hostiles_.reserve(kColumns_*kRows_);
+
+    scene_ = new QGraphicsScene(this);
+    view_ = new QGraphicsView(scene_, this);
+
+    int uiWidth = (kColumns_ + 4) * cellSize_;
+    int uiHeight = (kRows_ + 1) * cellSize_;
+
+    this->setFixedSize(uiWidth, uiHeight);
+    setCentralWidget(view_);
+
+    gameTimer_ = new QTimer(this);
+    gameTimer_->start(250);
+    connect(gameTimer_, SIGNAL(timeout()), this, SLOT(updateGameTime()));
```
Опять же, ничего не поменялось, в обоих конструкторах один и тот же код повторяется дважды... -> дубляж кода. Весь дубляж кода вынести в отдельный метод / методы и уже вызывать в конструкторах данный метод / методы.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+    scene_ = new QGraphicsScene(this);
+    view_ = new QGraphicsView(scene_, this);
+
+    int uiWidth = (kColumns_ + 4) * cellSize_;
+    int uiHeight = (kRows_ + 1) * cellSize_;
```
Опять 1 и 4 - числа, уже 3е исправление. Все это - магические числа. Как я уже писал, у вас 1, 4, 6, 8 встречаются числами во всем коде - все это магические числа. Исправьте во ВСЕМ проекте.

1 - не во вех местах маг. число, но там, где оно идет по логике с 4, 6, 8 - магическое точно.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+                for (int j = 0; j < kColumns_; ++j)
+                {
+                    gameGrid_[i][j] = gameGrid[i][j];
+                    if(gameGrid_[i][j] == 3)
+                    {
+                        p.x = j;
+                        p.y = i;
+                        newHostile.setPosition(p);
+                        qDebug() << "Hostile Coordinates: (" << p.x << ", " << p.y << ")";
+                        hostiles_.push_back(newHostile);
+                    }
+                    if(gameGrid_[i][j] == 4)
+                    {
+                        counsCount_++;
+                    }
+                    if(gameGrid_[i][j] == 2)
+                    {
+                        player_.setY(i);
+                        player_.setX(j);
+                    }
```
Пробелы после If

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+                        qDebug() << "Hostile Coordinates: (" << p.x << ", " << p.y << ")";
+                        hostiles_.push_back(newHostile);
+                    }
+                    if(gameGrid_[i][j] == 4)
+                    {
+                        counsCount_++;
+                    }
+                    if(gameGrid_[i][j] == 2)
```
4 и 2 и еще выше 3 - это что? Магические значения.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+    else
+    {
+        std::string str1 = "Game Over, coins left: ";
+        std::string str2 = std::to_string(counsCount_);
```
Str1 и Str2 - бесполезные имена, которые не несут никакого контекста. Было еще в первых замечаниях, до сих пор не исправлено.

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+
+    int element = 1;
+    int count = 5;
+    generateRandomElements(element,count);
+
+    element = 3;
+    count = 2;
+    generateRandomElements(element,count);
+
+    element = 4;
+    generateRandomElements(element,counsCount_);
```
Пробелы после запятой

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+void MainWindow::generateRandomElements(int element, int count)
+{
+    Hostile newHostile;
+    Point p(0,0);
```
Название с сокращением не исправляется уже 3ий раз!

![image](https://github.com/hBuzzy/Pacman-task/assets/47574560/03351990-6fdb-46f8-989f-976ef2027a0e)

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+            randomRow = std::rand() % kRows_;
+            randomColumn = std::rand() % kColumns_;
+        }
+        if(element == 3)
```
![image](https://github.com/hBuzzy/Pacman-task/assets/47574560/7af102ed-ca6b-4af3-aa7f-b030c1f15a3b)

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+        }
+        if (direction[0] == Directions::Left)
+        {
+            hostilePosition = currentHostile.getPosition();
+            gameGrid_[hostilePosition.y][hostilePosition.x] = currentHostile.getPreviousElement();
+            currentHostile.setPreviousElement(gameGrid_[hostilePosition.y][hostilePosition.x - 1]);
+            hostilePosition.x = hostilePosition.x - 1;
+            gameGrid_[hostilePosition.y][hostilePosition.x] = 3;
+            currentHostile.setPosition(hostilePosition);
+        }
+        if (direction[0] == Directions::Right)
+        {
+            hostilePosition = currentHostile.getPosition();
+            gameGrid_[hostilePosition.y][hostilePosition.x] = currentHostile.getPreviousElement();
+            currentHostile.setPreviousElement(gameGrid_[hostilePosition.y][hostilePosition.x + 1]);
+            hostilePosition.x = hostilePosition.x + 1;
+            gameGrid_[hostilePosition.y][hostilePosition.x] = 3;
+            currentHostile.setPosition(hostilePosition);
+        }
+        if (direction[0] == Directions::Up)
+        {
+
+            hostilePosition = currentHostile.getPosition();
+            gameGrid_[hostilePosition.y][hostilePosition.x] = currentHostile.getPreviousElement();
+            currentHostile.setPreviousElement(gameGrid_[hostilePosition.y-1][hostilePosition.x]);
+            hostilePosition.y = hostilePosition.y - 1;
+            gameGrid_[hostilePosition.y][hostilePosition.x] = 3;
+            qDebug() << "gameGrid_ Coordinates: (" << hostilePosition.x << ", " << hostilePosition.y << ")";
+            currentHostile.setPosition(hostilePosition);
+        }
+        if (direction[0] == Directions::Down)
+        {
+
+            hostilePosition = currentHostile.getPosition();
+            gameGrid_[hostilePosition.y][hostilePosition.x] = currentHostile.getPreviousElement();
+            currentHostile.setPreviousElement(gameGrid_[hostilePosition.y+1][hostilePosition.x]);
+            hostilePosition.y = hostilePosition.y + 1;
+            gameGrid_[hostilePosition.y][hostilePosition.x] = 3;
+            qDebug() << "gameGrid_ Coordinates: (" << hostilePosition.x << ", " << hostilePosition.y << ")";
+            currentHostile.setPosition(hostilePosition);
```
4 одинаковых куска кода, отличающиеся только параметрами -> Дубляж кода! Вынесите повторяющийся код в метод / методы!

hBuzzy commented on mainwindow.cpp:
```diff
@@ -1,15 +1,509 @@
+    }
+}
+
+void MainWindow::movePlayerRight()
```
Опять же, все методы движения - дубляж кода.... опять не исправлено. Весь повторяющийся код нужно разбивать и выносить отдельно...

hBuzzy commented on redactor.cpp:
```diff
@@ -0,0 +1,307 @@
+
+    connect(view, &CustomGraphicsView::customMouseRelease, this, &Redactor::handleCustomMouseRelease);
+
+    int uiWidth = (kColumns_ + 6) * cellSize_;
+    int uiHeight = (kRows_ + 1) * cellSize_;
```
6 и 1 все еще дубляж кода. Вообще у вас уже есть схожий конструктор  в классе mainwinow.... прямо напрашивается создать базовый класс и вынести эту функциональность в нее.

hBuzzy reviewed (changes requested):
Исправить замечания. Внимательно просмотрите все замечания, некоторые замечания не исправляются уже 3ий раз.

KloopRE committed:
Замечания

else if нужно перенести

Опять 1 и 4 - числа, уже 3е исправление. Все это - магические числа. Как я уже писал, у вас 1, 4, 6, 8 встречаются числами во всем коде - все это магические числа. Исправьте во ВСЕМ проекте.

1 - не во вех местах маг. число, но там, где оно идет по логике с 4, 6, 8 - магическое точно.

str1 ? Почему она тут 1? Дайте нормальное название.

Не исправлено сокращение

Название с сокращением не исправляется уже 3ий раз!

Пробел после if, скобку не переносим
player_.getY()+1 Пробелы.

6 и 1 все еще дубляж кода. Вообще у вас уже есть схожий конструктор в классе mainwinow.... прямо напрашивается создать базовый класс и вынести эту функциональность в нее.

str1 - бесполезное название, дайте осмысленное название.

hBuzzy reviewed (commented):
Эх, не буду больше вас тиранить, задачу приму.

Однако, в задаче все еще куча проблем. Самые важные из них:
1) Дубляж кода все так же и остался актуальным, а это очень плохо.
2) Скобки так же летают по проекту. Либо везде переносим, либо везде - нет. Точка. Не так, что для методов переносим, а для тела метода - нет.

В целом, работы проделано достаточно, если будет время и сможете исправить остаток замечаний, я даже думаю, что накину вам доп. баллы, к примеру, к паутине с 0.5 текущих до 1.
